### ISSUE TITLE
`vm start` won't run if `/challenge` is not accessible

### ISSUE THREAD

**Issue by ltfish** [pwncollege/dojo#248]
For some challenges (e.g., web challenges) we may make `/challenge` not accessible by the `hacker` user. However, the current logic in `10_start_vm.sh` will only invoke `vm start` if `/challenge` is accessible.

IMO we should fix the logic in `10_start_vm.sh` and avoid the unreliable auto-detection of which VM type to launch, which should also fix this issue.

**Comment by ConnorNelson** [comment]
`10_start_vm.sh` is only intended as a speedup mechanism. The user should be able to `vm start` whenever they want, or even just `vm connect`, which will also automatically start the vm if is not already started. When we see a `.ko` inside `/challenge`, we *know* the user will want to start the vm, and so that's what we do. 

If you're going to hide `/challenge`, then we have no way of knowing that we want to auto-start the vm.

If you want to automatically start the vm without a `.ko` file existing, we already have `/challenge/.init`, which will be automatically run at container start as the root user.

**Comment by Scoder12** [comment]
Connor, the script was recently modified in the windows PR to remove setuid from the script if the .ko file was not detected.

**Comment by Scoder12** [comment]
Oh wait we never actually went through with that

**Comment by ltfish** [comment]
> The user should be able to vm start whenever they want

So are you suggesting that we should always put `vm start` in `.init`? Or should we educate users (the players) that they should run `vm start` after starting a challenge?

The current logic basically is "if /challenge is not accessible by the hacker user, we don't start the VM." But /challenge can be inaccessible by `hacker` for legitimate reasons, e.g., it's a web challenge and `/challenge` should only be accessed by `www-data`. Not starting the VM seems weird to me.

> When we see a .ko inside /challenge, we know the user will want to start the vm, and so that's what we do.

How does it work when there are no `.ko` files in `/challenge`? Based on what you described, `vm start` should never be executed if `/challenge` is accessible but no `.ko` files exist inside. How does the VM start for other challenges where no `.ko` files exists?

**Comment by ltfish** [comment]
Sorry, I just realized that the real issue is that `find /challenge` will *fail* if `/challenge` is not accessible, which leads to `10_start_vm.sh` to fail, and no future starting scripts will run after that. The problem is actually not that `vm start` is not called. The problem is `10_start_vm.sh` does not ignore failures.

**Comment by ConnorNelson** [comment]
Is this causing other entrypoint scripts to not run? 

To be clear, the virtual machine is only required if your challenge *needs* a virtual machine. If you know your challenge *needs* a virtual machine, I think it makes sense for the challenge to start the virtual machine with a `vm start` inside `.init`. This really only saves ~5s of startup for the user. At some point, they're going to need to do a `vm connect` anyways to enter the virtual machine (Unless you're doing some networking to have services running in the virtual machine, but which are accessible from the container, I guess).

I wouldn't really expect that a web challenge requires the virtual machine, unless the container sandbox is too restrictive for something you're trying to do.

**Comment by ltfish** [comment]
> Is this causing other entrypoint scripts to not run?

Correct. My issue has nothing to do with the vm itself. It does need other entrypoint scripts to run so that the code server and no-vnc-server all run.
